prov/util/src/ipc_handle_monitor.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * (C) Copyright (c) 2022 Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_mr.h"
+#include "ofi_hmem.h"
+
+static int ipc_handle_monitor_subscribe(struct ofi_mem_monitor *monitor, const void *addr,
+					size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	/* no-op */
+	return FI_SUCCESS;
+}
+
+static void ipc_handle_monitor_unsubscribe(struct ofi_mem_monitor *monitor, const void *addr,
+					   size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	/* no-op */
+}
+
+static bool ipc_handle_monitor_valid(struct ofi_mem_monitor *monitor,
+			const struct ofi_mr_info *info, struct ofi_mr_entry *entry)
+{
+	return (memcmp((void **)&info->ipc_handle, (void **)&entry->info.ipc_handle,
+		IPC_HANDLE_SIZE) == 0) ? true : false;
```

Comment:
I am not sure if there is a better solution here. We could only compare the actual size of the IPC handle for different hmem ifaces. But that could require an extra variable in `ipc_info`.

Reply:
Can a single instance of an IPC cache realistically support multiple hmem interfaces?  We don't need the size of the ipc handle carried in every instance of ipc_info.  We only need the size per iface.  And there should probably be different monitors per iface.

Reply:
> Can a single instance of an IPC cache realistically support multiple hmem interfaces? We don't need the size of the ipc handle carried in every instance of ipc_info. 

I do not think a single instance of `ipc_cache` (`ofi_mr_cache`) should support  multiple hmem ifaces at the same time. The current design expects that only 1 hmem iface is being used in shm transfer.

> We only need the size per iface. And there should probably be different monitors per iface.

Currently the ipc_monitor is assigned to `ofi_mem_monitors[FI_HMEM_CUDA]` for cuda. 
```
struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
		[FI_HMEM_CUDA] = ipc_monitor,
	};
```

I agree we should be able to expand it to more hmem ifaces, but I thought they will just be something like

```
struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
		[FI_HMEM_CUDA] = ipc_monitor,
                [FI_HMEM_ZE] = ipc_monitor,
                [FI_HMEM_ROCR] = ipc_monitor,
                ....
	};
```

The current implementation of `ipc_monitor` in `prov/util/src/ipc_handle_monitor.c` is general and not depending on hmem iface. I think you want to make the compare size in `valid()` different for different hmem ifaces, does that mean we want to change `prov/util/src/ipc_handle_monitor.c` to `prov/util/src/cuda_ipc_handle_monitor.c` and create similar files for other hmem ifaces? Not sure if it's the best practice because it's hard for me to think there will be much difference for the ipc monitor implementation for other hmem ifaces besides the ipc handle size.